### PR TITLE
Auth Proxy: encoding of non-ASCII headers

### LINKED ISSUE #44744: Auth proxy to support non-ASCII characters
**What happened**:
User email and names may have non-ASCII characters. Since auth proxy uses HTTP headers to pass the user information, there is no good way to pass the correct non-ASCII characters.

**What you expected to happen**:
Find a way to pass non-ASCII characters (maybe encoded) in auth proxy, and Grafana can store and display it correctly (maybe decoded first)

**How to reproduce it (as minimally and precisely as possible)**:
When using auth proxy as auth provider, try to pass non-ASCII user name

**Anything else we need to know?**:

**Environment**:
- Grafana version: 8.3
- Data source type & version:
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by kostrse**
**What this PR does / why we need it**:

Information passed by auth proxy can contain non-ASCII characters (user names in native language,  email addresses containing Unicode characters etc.). Passing information via HTTP headers creates technical limitation, as headers cannot have non-ASCII characters.

HTTP headers used by auth proxy have to support encoding to avoid character set limitation.

**Which issue(s) this PR fixes**:

Fixes #44744

**Special notes for your reviewer**:

For backwards compatibility header values considered as encoded only if it's explicitly configured in settings:
```ini
[auth.proxy]
...
headers_encoded = true
```

As encoding format I chose quoted-printable (3) because Go doesn't have a parser for unicode escape sequences (2) and I would prefer to not write own parser for code which is responsible for authentication.
```
X-WEBAUTH-USER: M=C3=BCnchen
X-WEBAUTH-EMAIL: m=C3=BCnchen@=E5=85=AC=E5=8F=B8.com
X-WEBAUTH-ROLE: =CE=94=CE=B9=CE=B1=CF=87=CE=B5=CE=B9=CF=81=CE=B9=CF=83=CF=84=CE=AE=CF=82
```

### CODE REVIEW

**Comment by sdboyer** [comment]
I'm amenable in general, but would prefer the config approach (option 2 that you mentioned in #44744) in order to guard against possible problems. The default should be to use the existing logic that does not decode.

File: pkg/services/contexthandler/authproxy/authproxy.go [hunk 1]
```diff
@@ -84,14 +84,19 @@ type Options struct {
...
-	header := options.Ctx.Req.Header.Get(cfg.AuthProxyHeaderName)
-	return &AuthProxy{
+	auth := &AuthProxy{
 		remoteCache: options.RemoteCache,
 		cfg:         cfg,
 		ctx:         options.Ctx,
 		orgID:       options.OrgID,
-		header:      header,
 	}
+	auth.init()
```
Comment (sdboyer) [inline pkg/services/contexthandler/authproxy/authproxy.go:93]: nit-ish: What benefit do we get from introducing a separate `init()` method? Why not do this directly in `New()`?

**Review by sdboyer — COMMENTED** [review]
Apologies for the delayed review.

Generally, this LGTM. I have what's mostly just a nit - once we resolve that discussion, i think we can merge.

File: pkg/services/contexthandler/authproxy/authproxy.go (hunk 1 above)
Comment (kostrse) [reply pkg/services/contexthandler/authproxy/authproxy.go:93]: Agree, not needed. Removed.

**Review by sdboyer — APPROVED** [review]
LGTM!

**Pull request merged**